lib/logstash/inputs/gelf.rb

```diff
@@ -267,11 +267,25 @@ def remap_gelf(event)
   end
 
   def strip_leading_underscore(event)
-     # Map all '_foo' fields to simply 'foo'
-     event.to_hash.keys.each do |key|
-       next unless key[0,1] == "_"
-       event.set(key[1..-1], event.get(key))
-       event.remove(key)
-     end
+    # Map all '_foo' fields to simply 'foo'
+    event.to_hash.keys.each do |key|
+      move_field(event, key, key.slice(1..-1)) if key.start_with?('_')
+    end
+  end
+
+  def move_field(event, source_field, destination_field)
+    value = event.get(source_field)
+    value = coerce_timestamp_carefully(value) if destination_field == LogStash::Event::TIMESTAMP
+    event.set(destination_field, value)
+    event.remove(source_field)
+  rescue => e
+    @logger.warn("Failed to move field `#{source_field}` to `#{destination_field}`: #{e.message}")
+    event.tag("_gelf_move_field_failure")
+  end
+
+  def coerce_timestamp_carefully(value)
+    # catch float numbers in 123.567 or 0.123567e3 forms
+    value = BigDecimal.new(value) if value.kind_of?(String) && value.match(/\A([0-9]+(\.[0-9]+)?)|(0\.[0-9]+e[0-9])\z/)
```

Comment:
Without this change the test at https://github.com/logstash-plugins/logstash-input-gelf/pull/67/files#diff-977179e372429b2780ffe4f8e97ed783160981021b1d4b3f92aa17bd40d31906R109 fails.
The reason is due to fact that the:
```ruby
gelfclient.notify!("short_message" => "msg1", "_@timestamp" => BigDecimal.new("946702800.1"))
```
generate a GELF message where the BigDecimal is converted into the String value `0.9467028001e9` that's what `BigDecimal.new("946702800.1").to_s` return.

So at that point the String `0.9467028001e9` passed into  `LogStash::Timestamp.at` raises an error that make the event to be tagged and the `@timestamp` bring value "now", causing the test to fail.

So I think we should be prepared to accept floats in scientific notation strings or we could leave that part as before:
```ruby
  def coerce_timestamp_carefully(value)
    LogStash::Timestamp.at(value)
  rescue TypeError => e
    # maybe it's a BigDecimal?
    coerced_value = BigDecimal.new(value)
    LogStash::Timestamp.at(coerced_value.to_i, coerced_value.frac * 1000000)
  end
```

giving the responsibility of correct parsing to the `BigDecimal` constructor.

Reply:
I would still like to avoid reacting to exceptions thrown by `LogStash::Timestamp::at(Object)`, since exceptions aren't cheap at scale.

We know that `LogStash::Timestamp::at(String)` will always fail, so if we encounter a `String`, passing it to `Kernel::BigDecimal(String)` at least won't make things _worse_; in this way we can defer to BigDecimal for the parsing without having to map every possible string format of a decimal:

```suggestion
    value = BigDecimal(value) if value.kind_of?(String)
```